File: javascript/typescript/inventory.ts
class InventoryItem {
  id: string;
  name: string;
  price: number;

  constructor(id: string, name: string, price: number) {
    this.id = id;
    this.name = name;
    this.price = price;
  }
}

enum StoreEventType {
  PURCHASED,
  REFUNDED
}

class StoreEvent {
  eventType: StoreEventType;
  timestamp = Date.now();
  payload: unknown;

  constructor(eventType: number, payload: unknown) {
    this.eventType = eventType;
    this.payload = payload;
  }
}

class Store {
  name: string;
  openingTime = 900;
  closingTime = 2100;
  stock: Record<string, number> = {};
  events: StoreEvent[] = [];

  constructor(name: string) {
    this.name = name;
  }

  updateItemQuantity(item: InventoryItem, amount: number) {
    this.stock[item.id] = amount;
  }

  purchaseItem(item: InventoryItem) {
    const currentItemStock = this.stock[item.id] || 0;
    if (currentItemStock <= 0) return;
    this.events.push(new StoreEvent(StoreEventType.PURCHASED, item.id));
    this.updateItemQuantity(item, this.stock[item.id] - 1);
  }

  refundItem(item: InventoryItem) {
    this.events.push(new StoreEvent(StoreEventType.REFUNDED, item.id));
    this.updateItemQuantity(item, this.stock[item.id] + 1);
  }
}

const item1 = new InventoryItem("1", "Item 1 - A", 500);
const item2 = new InventoryItem("2", "Item 2 - B", 750);
const item3 = new InventoryItem("3", "Item 3 - C", 1000);

const store = new Store("Bob's Shop");

store.updateItemQuantity(item1, 10);
store.updateItemQuantity(item2, 7);
store.updateItemQuantity(item3, 5);

store.purchaseItem(item1);

console.log(store.stock);

File: python/inventory.py
from enum import Enum
from datetime import datetime
from dataclasses import dataclass, field
from typing import List, Dict

@dataclass
class InventoryItem:
    id: str
    name: str
    price: int

class StoreEventType(Enum):
    PURCHASED = 1
    REFUNDED = 2

@dataclass
class StoreEvent:
    event_type: StoreEventType
    timestamp: datetime = datetime.now()
    payload: str = ""

@dataclass
class Store:
    name: str
    opening_time: datetime = datetime.now()
    closing_time: datetime = datetime.now()
    stock: Dict[str, int] = field(default_factory = dict)
    events: List[StoreEvent] = field(default_factory = list)

    def update_item_quantity(self, item, amount):
        self.stock[item.id] = amount

    def purchase_item(self, item):
        if self.stock[item.id] < 1:
            return
        self.events.append(StoreEvent(event_type = StoreEventType.PURCHASED, payload = item.id))
        self.update_item_quantity(item, self.stock[item.id] - 1)

    def refund_item(self, item):
        self.events.append(StoreEvent(event_type = StoreEventType.REFUNDED, payload = item.id))
        self.update_item_quantity(item, self.stock[item.id] + 1)

item1 = InventoryItem(id = "1", name = "Item 1 - A", price = 500)
item2 = InventoryItem(id = "2", name = "Item 2 - B", price = 750)
item3 = InventoryItem(id = "3", name = "Item 3 - C", price = 1000)

store = Store(name = "Bob's Shop");

store.update_item_quantity(item1, 10);
store.update_item_quantity(item2, 7);
store.update_item_quantity(item3, 5);

store.purchase_item(item1);

print(store.stock);
